DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,617 @@
+    immutable int* p1 = systemPtr; // error, reading @system var of unsafe type
+    int i0 = safeInt;              // allowed
+    int i1 = systemInt;            // allowed, not an unsafe type
+}
+```
+
+**(3) Variables and fields without annotation are `@safe` unless their initial value is not `@safe`**
+
+The rules regarding variables and fields are as follows:
+- An initialization expression `x` is `@system` when the function `(() => x)` is inferred as `@system`.
+- When marked `@system`, the result is always `@system` regardless of the type.
+- When marked `@trusted`, the initialization expression `x` is treated as `(() @trusted => x)`.
+- When marked `@safe`, the initialization expression must be `@safe`.
+- In the absence of an annotation, the result is `@system` only if the type is unsafe and the initialization expression is `@system`.
+
+```D
+int* getPtr() @system {return cast(int*) 0x8035FDF0;}
+int  getVal() @system {return -1;}
+
+extern int* x0;                   // @safe by default
+int* x1 = x0;                     // @safe, (() => x0) is @safe
+int* x2 = cast(int*) 0x8035FDF0;  // @system, (() => cast(int*) 0x8035FDF0) is @system
+int* x3 = getPtr();               // @system, (() => getPtr()) is @system
+int  x4 = getVal();               // @safe, int is not an unsafe type
+@system int x5 = 1;               // @system as requested
+@trusted int* x6 = getPtr();      // @safe, the getPtr call gets trusted
+@safe int* x7 = getPtr();         // error: cannot initialize @safe variable with @system initializer
+
+struct S {
+    // same rules for fields:
+    int* x9 = x3; // @system
+    int  x8 = x5; // @safe
+}
+```
+
+An exception to the last rules is made on unsafe types when the compiler knows the resulting value is safe.
+```D
+int* getNull() pure @system {return null;}
+int* n = getNull(); // despite unsafe type with @system initialization expression, inferred as @safe
+```
+
+Annotations with a scope (`@system {}`) or colon (`@system:`) affect variables just like they do functions.
+```D
+@system {
+    int y0; // @system
+}
+
+@system:
+int y1; // @system
+```
+
+N.B. functions are currently `@system` by default, but that is planned to be changed: [Make @safe the default](https://github.com/dlang/DIPs/pull/166)
+
+**(4) `__traits(getFunctionAttributes)` may be called on variables and fields**
+
+Currently it is possible to give function attributes to declarations that aren't functions.
+It is not possible however to inspect any of them.
+
+```D
+@system @nogc pure nothrow int x;
+pragma(msg, __traits(getFunctionAttributes, x)); // error: first argument is not a function
+pragma(msg, __traits(getAttributes, x)); // tuple()
+```
+
+Since memory safety-related attributes now have an effect on variables and fields, it becomes useful to inspect them.
+Therefor the restriction on the getFunctionAttributes trait gets lifted.
+
+The name "function attributes" is a bit unfortunate in this case, but this DIP does not aim to fix that.
```

Comment:
``getDeclarationAttributes``
Add a second name, deprecate old one. This DIP can and probably should fix this (traits don't need a DIP to be added).

Reply:
While the list of traits has gotten bigger without much resistance, I am not a big fan of that direction. It's messy right now, and Atila mentioned how he wanted a more uniform way for introspection, so I try to avoid touching the pile traits right now. Thanks for your suggestion though, if I get some more opinions on this I'll try to accommodate.